### DISCUSSION TITLE
Default 'False' output of 0 for Or components

### DISCUSSION THREAD

**Discussion by LeakingAmps** [FakeFishGames/Barotrauma#9866] [Ideas and suggestions]
I just spent a few hours trying to figure out why my circuits weren't working, until I finally discovered that the Or gate (by default) only functions like half an or gate.

I'm sure there's reasons to have it able to do this, but I can't see any good reason to make wiring _more_ confusing by not making it have a default false output of 0 (and that's not including that if you want to use it a bunch, you'll likely have to set the default output on each one, which get quite tedious)

If there's a way to change default component settings on my client, that'd be wonderful, but I don't see a way to do that either.

Edit: Just discovered that And gates do the same thing. Please?

**Comment by mygamingaccount** [comment]
A good point for having it like that is that gates can shut down when there's no false output, so the result will be more optimal with fewer CPU cycles calculating the logic outputs.

However all of the vanilla ships that _could_ make use of this _don't_. You see loads of OR components chained up with their false outputs set to `0` when they could've been left empty.

There is an optimization feature that not even the original developers make use of. Then why keep it around?

**Reply by LeakingAmps** [reply]
I'm not asking them to remove this feature - I recognize that it could be *really* helpful behavior if you want it to be there, I'm just asking for a change to the default behavior so it's more intuitive to people new to the system that has no ingame tutorials, a threadbare wiki, and few up-to-date tutorials.

**Reply by gerakis** [reply]
Mmmm you still have to evaluate whether the input changed. You cannot shut them down. Evaluating something as a null is still an evaluation.

Evaluating a statement only takes a single cpu cycle. Its considered insanely fast.

**Reply by LeakingAmps** [reply]
I believe MyGamingAccount is assuming the circuit network is update-based, and only calculated if it receives an update. This *would* mean that having gates with no false output could cause fewer updates down the line, as they would only propagate updates if they output true.

In scenario 1, a lever is hooked up to an or gate, leading into a large tangled mess of gates. If false output is defined, flipping the lever off and turning off the gate also sends an update into the mass of gates.

In Scenario 2, the same settup, but the false output is undefined. In this case, flipping the lever off just updates the first or gate and nothing else.

I have no idea if this is how circuits work in this game though, and I suspect they would, as you pointed out, have very light loads on the CPU, as they're math/string operations, and nothing too fancy.

**Reply by FalcoGer** [reply]
@LeakingAmps But when you turn the lever off you want something to happen. That is exactly the problem with not defining the output to be 0. You WANT the circuit to update. Otherwise what's the point? You hook up your circuit to something to set state. and you can only set the state one way. Even then, the whole mess of the circuit still updates, it just sets everything to null signals, which are useless, unless you want another signal to come into the same inputs of some gates.

Again, they are situationally useful, but it should be set explicitly if you want them. 95% of the time, I just want normal logic and half the time I forget to set the false output, because it makes no sense for it to be empty in the first place.

**Reply by mygamingaccount** [reply]
https://github.com/Regalis11/Barotrauma/blob/2efdead44bf2fd76adebe19ee881d3c1eb147455/Barotrauma/BarotraumaShared/SharedSource/Items/Components/Signal/BooleanOperatorComponent/BooleanOperatorComponent.cs#L82-L101

**Reply by mygamingaccount** [reply]
Components stop receiving updates when they are inactive

**Reply by mygamingaccount** [reply]
@FalcoGer null signals are equivalent to `0` when it comes to logic component inputs. Those only look for nonzero signals.

**Reply by LeakingAmps** [reply]
From what I've seen, that's not the case @mygamingaccount 

Specifically, several or gate with 1/null as true/false outputs chained together, hooked up to a lamp at the end, the lamp will be in its default state until any or gate receives a signal, but after that, the signal from the last or gate will stay high until the ship resets.

This is useful for some cases, but in the general case it makes things not work as expected. I'm not asking them to remove null signals, since I'm sure there are good use cases. However, I'd like to at least be able to change the default output from null to 0 on components on my client, if not as the default setting for new users. Default settings should be set up for those new to the game/designer, not for those that already know their way around it.

**Reply by mygamingaccount** [reply]
Correct. The lamp needs a 0 signal to reset. The logic components don't.

The default should be false set to `0`, but also the vanilla ships should demonstrate how to optimally use components, as they are the first step new players use for learning how systems work. Default subs should therefore change their components to make use of the optimization feature described above.

And if this causes performance issues with too many updates, then solve that by not updating components in storage that aren't attached to walls.

**Comment by mygamingaccount** [comment]
You seem like the kind of engineer who makes a design with open-collector gates then forgets the pull-up resistors

**Reply by LeakingAmps** [reply]
I went to college for programming, not ECE, so I don't know what that means, beyond that it's probably an insult?

**Reply by mygamingaccount** [reply]
Ha! No, not an insult. Now you know what it feels like doing what I wrote earlier. It's a rite of passage.
You were supposed to know since it's all written in the manual that doesn't exist.

**Reply by gerakis** [reply]
It was totally an insult.

While I get that you had a bad time, other people would also have a bad time if the situation were reversed. Someone will always want it the other way. Eventually you just get used to double checking.

**Comment by FalcoGer** [comment]
I agree. Any logic gates should start with 0 on the false output, that goes for and, or, xor, greater than, equals and so on. Whenever I place one of those components down, there is a 50% chance i forget to put that 0 in there, because it makes no sense to just have it empty in my head. Null signals are terrible idea in the first place. And after that, i test my circuit and it doesn't work and then i have to click every gate and check the outputs, it's just annoying. If you want a null signal you should opt into that, not out. The definition of OR is that the output is 0 if all inputs are 0, otherwise the output is 1. The way it is by default just isn't like that, and it trips me up every time.

**Reply by mygamingaccount** [reply]
Following your definition wouldn't be very helpful. It wouldn't be possible to have many buttons on one OR component for example. It's more useful the way it currently is:

https://github.com/Regalis11/Barotrauma/blob/master/Barotrauma/BarotraumaShared/SharedSource/Items/Components/Signal/BooleanOperatorComponent/OrComponent.cs

In other words: If any of the inputs are nonzero, output is true.

**Reply by gerakis** [reply]
I just want to say I use null signals constantly. Its very useful for a lot of builds.

**Reply by FalcoGer** [reply]
The definition of or: the output is 1 if any inputs are high. A floating input is indeterminate. If it picks up even the slightest EMI or be capacitively coupled to something it will flicker high inconsistently. Needless to say that you don't want that. Leaving your logic gate inputs floating is ridiculous in electrical engineering. buttons that are not pressed are 0, not floating. Light bulbs may be left floating if you flick the switch off, because they are driven by power and thus stop working if you cut the power. But logic circuits are always powered, even if you don't supply any inputs. That is why you use a resistor to force the terminals either direction, not leave them floating, because logic levels are determined by voltage, not power. And voltage is just a potential and you can have any voltage on a floating connection and you can't be sure what it is. Wiring inputs to the same terminal is also ridiculous, that's just a short circuit waiting to happen. Best case you loose some power, worst case you fry your components or start a fire.

**Reply by mygamingaccount** [reply]
That's not the definition of the OR component, it's

"Sends a signal if either of the inputs receives a signal"

"Floating input" is not indeterminate, it is equal to 0

There is no EMI in Barotrauma.

Stop pretending that you're an engineer

**Reply by FalcoGer** [reply]
I tinkered with electronics long enough to know what I'm talking about. A floating signal can float high, which would make it output. The truth table for an OR is indeterminate for a floating input.

Literally take a digital voltmeter and stick the probes in the air. It will not be 0. Surely *your* engineering degree didn't prepare you for that. I would link you to a few articles that pop up when searching "floating inputs on logic circuits" on your favorite search engine, but I think an engineer like you can figure out how to do that. Or maybe you could just put some ICs on a breadboard and do the experiment yourself. But you'd rather be complaining here about a non-issue, because you're a troll.

I say the game should reflect reality, not the game doesn't work like it says it does. I say that because the way it works like it does is not correct, confusing, illogical and needlessly complicated.

Also please stop trolling. Your constant "No, that's stupid, we don't need that in the game" regardless of what the idea is, just because you got used to the way the game is or because your big brain can't think of a use case is irritating.

I'm not even proposing that you get rid of floating inputs anyway. I just say the default state should be 0, because that is what every computer science student learns and they learn it that way because that is the way it is. What the heck is the problem with that?

If you have a valid argument to make, then argue the point, not insult the people who make the arguments or spout falsehoods that are easily proven to be the garbage that they are.

The ACTUAL definition of an or gate:
https://en.wikipedia.org/wiki/OR_gate

> The OR gate is a digital logic gate that implements logical disjunction (v) from mathematical logic – it behaves according to the truth table above [which does not contain a floating input]. A HIGH output (1) results if one or both the inputs to the gate are HIGH (1). If neither input is high, a LOW output (0) results.

The ACTUAL behavior of floating inputs (hint: they are not 0):
https://www.microchip.com/forums/m447764.aspx

> If an input is floating, it will be "high impedance". Therefore it will be flipping up and down at random, perhaps picking up stray RF etc. So tie the pin to ground or rail and it can't do that. Leaving an input floating is a recipe for much unhappiness.

Please stop lying, please stop insulting people, please stop being a dick. Thank you.
I may not have a degree in electrical engineering, but I do in computer science and electronics is a hobby of mine. And unlike you, I know what an OR gate or a floating input is. And even if I didn't, I wouldn't assert things that I pull out of my ass as the truth without even the most basic of google searches.

**Reply by mygamingaccount** [reply]
Why don't you go and learn a few things before going around insulting people who are trying to educate you? Learn about open collectors for a start. You're trying to push your half-baked understanding of logic circuits into a literal video game.

You not being able to take criticism doesn't add any merit to your suggestions.

**Reply by FalcoGer** [reply]
I am providing sources while you seem to pull your "education" out of your rear end. Show me the evidence that a floating input is equivalent to 0 as you claim. And when all the sources that I find indicate that something is the way I learned and I experienced it that exact way myself through experimentation, then some guy telling me it's different without any sources, clearly I trust myself and the thousands of others that know how erratic a floating input on a circuit behaves instead of the one guy who thinks that not pulling things to ground is the same as a logical 0. There is a reason why people pull the inputs to ground if they want them off, and they wouldn't do that if you could just flip a switch off and have it floating.

Literally 10 seconds on google gave me this. Why you can't do that and stubbornly claim falsehoods is beyond me.
https://www.youtube.com/watch?v=i0mwHTYeEPI

you want to skip to 11 minutes.

**Reply by mygamingaccount** [reply]
The "floating" input on logic components is always equivalent to "0", Check the source of the base class: BooleanOperatorComponent.cs
The incoming signal of a value "0" is equivalent to no incoming signal at all, since it returns if immediately when this is the case.

This is very **very** useful since it means you don't need to have a million wires going to every single unused pin of all the components on your ship. You can not seriously be suggesting this to be changed, it would make it almost impossible for anyone without many hours of training to make new subs.
